### PR TITLE
sys-kernel/zen-sources: add 6.5.5, fix SRC_URI, add patch for Makefile

### PR DESCRIPTION
**Pull request by Fuchskaute**
The zen-sources package in the main tree is outdated by several months (version 6.3.9).

The SRC_URI has also changed, it seems zen-kernel changed their naming and extension from `-v${PV}-zen1.patch.xz` to `linux-v${PV}-zen1.patch.zst`.

I also included a patch for the Makefile, as it incorrectly sets the kernel version to 6.5.0 on kernel version 6.5.5. 

If there is a better way to handle these changes I would appreciate feedback.

### CODE REVIEW

**Comment by heftig** [comment]
Note that the `.zst` patches no longer include the stable patches and need to be applied to the matching stable release (v6.5.5 in this case), not the mainline release (v6.5). You're probably seeing the version difference in the Makefile because you're missing the stable patch.

**Comment by Fuchskaute** [comment]
Thank you for the quick reply and info, I'll take a look at it

**Comment by Fuchskaute** [comment]
I'm going to ask for guidance on #gentoo-kernel, I need to figure out how to properly apply the stable patches/genpatches *before* the ebuild tries  to  apply zen sources.

**Comment by kzdixon** [comment]
I just got around this in my ebuild by manually calling `zstd` for the patch... The normal portage unpacker doesn't seem to like `patch.zst` and I saw 0 results when searching the entire `gentoo` tree for any `patch.zst` examples.
```
# Copyright 1999-2022 Gentoo Authors
# Distributed under the terms of the GNU General Public License v2

EAPI="8"
ETYPE="sources"
K_WANT_GENPATCHES="base extras"
K_GENPATCHES_VER="7"
K_SECURITY_UNSUPPORTED="1"
K_NOSETEXTRAVERSION="1"

inherit kernel-2
detect_version
detect_arch

KEYWORDS="~amd64 ~arm64 ~x86"
HOMEPAGE="https://github.com/zen-kernel"
IUSE=""

DESCRIPTION="The Zen Kernel Live Sources"

ZEN_URI="https://github.com/zen-kernel/zen-kernel/releases/download/v${PV%_*}-zen${PV#*p}/linux-v${PV%_*}-zen${PV#*p}.patch.zst"
SRC_URI="${KERNEL_URI} ${GENPATCHES_URI} ${ARCH_URI} ${ZEN_URI}"

#UNIPATCH_LIST="${DISTDIR}/linux-v${PV%_*}-zen${PV#*p}.patch.zst"
UNIPATCH_LIST="${WORKDIR}/linux-v${PV%_*}-zen${PV#*p}.patch"
UNIPATCH_STRICTORDER="yes"

K_EXTRAEINFO="For more info on zen-sources, and for how to report problems, see: \
${HOMEPAGE}, also go to #zen-sources on oftc"

#src_unpack() {
#   zstd -df "/var/cache/distfiles/linux-v${PV%_*}-zen${PV#*p}.patch.zst" -o "${WORKDIR}/linux-v${PV%_*}-zen${PV#*p}.patch"
#}

pkg_setup() {
    ewarn
    ewarn "${PN} is *not* supported by the Gentoo Kernel Project in any way."
    ewarn "If you need support, please contact the zen developers directly."
    ewarn "Do *not* open bugs in Gentoo's bugzilla unless you have issues with"
    ewarn "the ebuilds. Thank you."
    ewarn
    kernel-2_pkg_setup
    # needed because of .zst change from .xz upstream; portage kernel-2 eclass doesn't seem to handle it and neither does normal unpack
    zstd -df "/var/cache/distfiles/linux-v${PV%_*}-zen${PV#*p}.patch.zst" -o "${WORKDIR}/linux-v${PV%_*}-zen${PV#*p}.patch"
}

pkg_postrm() {
    kernel-2_pkg_postrm
}
```
Correctly applied the patches for me finally.

**Comment by Fuchskaute** [comment]
Hello and thank you for helping me with this! 

When I copy your ebuild pkgdev manifest isn't able to source the ZEN_URI, after replacing it (and the filename for zstd unpack) I was able to reproduce your success!!

Output of `ebuild zen-sources-6.5.5.ebuild unpack`
https://0x0.st/HVuv.ebuild%20u

**Comment by Fuchskaute** [comment]
Would a good way to go about this be to submit the now working ebuild (with adjusted ZEN_URI) with you as a Co-author?

**Comment by kzdixon** [comment]
> Would a good way to go about this be to submit the now working ebuild (with adjusted ZEN_URI) with you as a Co-author?
> 
> I will push another commit with your changes and leave space in the commit message for you as a co-author which I will update if that's okay.

Just a heads up that my ebuild version relies on a version scheme of `X.X.X-pY -> X.X.X-zenY` which doesn't match the `gentoo` tree ebuild. I only did it that way to avoid having to go in and edit the zenX version manually whenever it had a release in that fashion. So you'd want to remove the slight edits I made there probably.

File: sys-kernel/zen-sources/zen-sources-6.5.5.ebuild [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+# Needed for zstd compression of the patch
+BDEPEND="app-arch/zstd"
+
+DESCRIPTION="The Zen Kernel Live Sources"
+
+ZEN_URI="https://github.com/zen-kernel/zen-kernel/releases/download/v${PV}-zen1/linux-v${PV}-zen1.patch.zst"
+SRC_URI="${KERNEL_URI} ${GENPATCHES_URI} ${ARCH_URI} ${ZEN_URI}"
+
+UNIPATCH_LIST="${WORKDIR}/linux-v${PV%_*}-zen${PV#*p}.patch"
```
Comment (kzdixon) [inline sys-kernel/zen-sources/zen-sources-6.5.5.ebuild:27]: `gentoo` ebuild does not follow my `-pY` version scheme.
```suggestion
UNIPATCH_LIST="${WORKDIR}/linux-v${PV}-zen1.patch"
```

File: sys-kernel/zen-sources/zen-sources-6.5.5.ebuild [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+UNIPATCH_STRICTORDER="yes"
+
+K_EXTRAEINFO="For more info on zen-sources, and for how to report problems, see: \
+${HOMEPAGE}, also go to #zen-sources on oftc"
+
+src_unpack() {
+	zstd -df "${DISTDIR}/linux-v${PV}-zen1.patch.zst" -o "${WORKDIR}/linux-v${PV%_*}-zen${PV#*p}.patch"
```
Comment (kzdixon) [inline sys-kernel/zen-sources/zen-sources-6.5.5.ebuild:34]: `gentoo` ebuild does not follow my `-pY` version scheme.
```suggestion
	zstd -df "${DISTDIR}/linux-v${PV}-zen1.patch.zst" -o "${WORKDIR}/linux-v${PV}-zen1.patch"
```

File: sys-kernel/zen-sources/zen-sources-6.5.5.ebuild [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+pkg_setup() {
+	ewarn
+	ewarn "${PN} is *not* supported by the Gentoo Kernel Project in any way."
+	ewarn "If you need support, please contact the zen developers directly."
+	ewarn "Do *not* open bugs in Gentoo's bugzilla unless you have issues with"
+	ewarn "the ebuilds. Thank you."
+	ewarn
+	kernel-2_pkg_setup
+}
+
```
Comment (kzdixon) [inline sys-kernel/zen-sources/zen-sources-6.5.5.ebuild:47]: Need to cleanup patch file or else it gets installed to `/usr/src`. There may be a better practice than `rm`.
```suggestion
src_install() {
    rm "${WORKDIR}/linux-v${PV}-zen1.patch"
    kernel-2_src_install
}
```

**Review by kzdixon — CHANGES_REQUESTED** [review]
Here are some changes that should leave it working with how the `gentoo` ebuild versions have been handled currently.

**Comment by Fuchskaute** [comment]
Thank you so much for the detailed examples and explanation

**Comment by Fuchskaute** [comment]
I have revised the ebuild a bit further, using the unpacker eclass to unpack the zst patch

File: sys-kernel/zen-sources/zen-sources-6.5.5.ebuild [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+detect_version
+detect_arch
+
+KEYWORDS="~amd64 ~arm64 ~x86"
+HOMEPAGE="https://github.com/zen-kernel"
+IUSE=""
+
+# Needed for zstd compression of the patch
+# Is this actually needed with the unpacker eclass?
+BDEPEND="app-arch/zstd"
```
Comment (thesamesam) [inline sys-kernel/zen-sources/zen-sources-6.5.5.ebuild:21]: To answer your q, not if we do:
```suggestion
BDEPEND="$(unpacker_src_uri_depends)"
```

File: sys-kernel/zen-sources/zen-sources-6.5.5.ebuild [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+	ewarn "${PN} is *not* supported by the Gentoo Kernel Project in any way."
+	ewarn "If you need support, please contact the zen developers directly."
+	ewarn "Do *not* open bugs in Gentoo's bugzilla unless you have issues with"
+	ewarn "the ebuilds. Thank you."
+	ewarn
+	kernel-2_pkg_setup
+}
+
+src_install() {
+	rm "${WORKDIR}/linux-v${PV}-zen1.patch"
```
Comment (thesamesam) [inline sys-kernel/zen-sources/zen-sources-6.5.5.ebuild:50]:
```suggestion
	rm "${WORKDIR}/linux-v${PV}-zen1.patch" || die
```

**Comment by thesamesam** [comment]
Merged in 70b88aa02daaaa76d02a85b90116b203b1607f18 (lost 'Closes:' b/c squashed)

**Pull request closed without merging**